x/accounts/vesting/vesting.go

```diff
@@ -0,0 +1,493 @@
+	Owner            collections.Item[[]byte]
+	OriginalVesting  collections.Map[string, math.Int]
+	DelegatedFree    collections.Map[string, math.Int]
+	DelegatedVesting collections.Map[string, math.Int]
+	addressCodec     address.Codec
+	headerService    header.Service
+	// Vesting end time, as unix timestamp (in seconds).
+	EndTime collections.Item[time.Time]
+}
+
+func (bva *BaseVesting) Init(ctx context.Context, msg *vestingtypes.MsgInitVestingAccount) (
+	*vestingtypes.MsgInitVestingAccountResponse, error,
+) {
+	owner, err := bva.addressCodec.StringToBytes(msg.Owner)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid 'owner' address: %s", err)
+	}
+	err = bva.Owner.Set(ctx, owner)
+	if err != nil {
+		return nil, err
+	}
+
+	funds := accountstd.Funds(ctx)
+	if !funds.Equal(msg.Amount) {
+		return nil, sdkerrors.ErrInvalidRequest.Wrap("invalid funding amount, should be equal to vesting amount")
+	}
+
+	if err := validateAmount(msg.Amount); err != nil {
+		return nil, err
+	}
+
+	sortedAmt := msg.Amount.Sort()
+	for _, coin := range sortedAmt {
+		err = bva.OriginalVesting.Set(ctx, coin.Denom, coin.Amount)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = bva.EndTime.Set(ctx, msg.EndTime)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vestingtypes.MsgInitVestingAccountResponse{}, nil
+}
+
+// ExecuteMessages handle the execution of codectypes Any messages
+// and update the vesting account DelegatedFree and DelegatedVesting
+// when delegate or undelegate is trigger. And check for locked coins
+// when performing a send message.
+func (bva *BaseVesting) ExecuteMessages(
+	ctx context.Context, msg *account_abstractionv1.MsgExecute, getVestingFunc getVestingFunc,
+) (
+	*account_abstractionv1.MsgExecuteResponse, error,
+) {
+	// we always want to ensure that this is called by the x/accounts module, it's the only trusted entrypoint.
+	// if we do not do this check then someone could call this method directly and bypass the authentication.
+	if !accountstd.SenderIsAccountsModule(ctx) {
+		return nil, fmt.Errorf("sender is not the x/accounts module")
+	}
+	owner, err := bva.Owner.Get(ctx)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid owner address: %s", err.Error())
+	}
+	bundler, err := bva.addressCodec.StringToBytes(msg.Bundler)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid bundler address: %s", err.Error())
+	}
+	if !bytes.Equal(owner, bundler) {
+		return nil, fmt.Errorf("bundler is not the owner of this vesting account")
+	}
+	hs := bva.headerService.GetHeaderInfo(ctx)
+
+	for _, m := range msg.ExecutionMessages {
+		concreteMsg, err := vestingtypes.UnpackAnyRaw(m)
+		if err != nil {
+			return nil, err
+		}
+
+		typeUrl := sdk.MsgTypeURL(concreteMsg)
+		switch typeUrl {
+		case MSG_DELEGATE:
+			msgDelegate, ok := concreteMsg.(*stakingtypes.MsgDelegate)
+			if !ok {
+				return nil, fmt.Errorf("invalid proto msg for type: %s", typeUrl)
```

Comment:
It seems unlikely that these errors will be triggered, but if they are, it'd be really handy to know the type of the `concreteMsg`. That way, if one of these errors does crop up, it's easier to identify which `Msg` needs to be fixed.

```suggestion
				return nil, fmt.Errorf("invalid proto msg %T for type: %s", concreteMsg, typeUrl)
```